### ISSUE TITLE
os.path.realpath returns invalid path for junction pointing to letter-less volume

### ISSUE THREAD

**Issue by 22ed5815-795f-4f8b-9a5a-1cb83ae53f75** [python/cpython#89760]
BPO | [45597](https://bugs.python.org/issue45597)
--- | :---
Nosy | @pfmoore, @tjguk, @zware, @eryksun, @zooba, @grv87

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-10-24.09:38:45.590>
labels = ['3.10', '3.8', '3.9', 'OS-windows']
title = 'os.path.realpath returns invalid path for junction pointing to letter-less volume'
updated_at = <Date 2021-10-24.11:15:13.756>
user = 'https://github.com/grv87'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-24.11:15:13.756>
actor = 'eryksun'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Windows']
creation = <Date 2021-10-24.09:38:45.590>
creator = 'grv87'
dependencies = []
files = []
hgrepos = []
issue_num = 45597
keywords = []
message_count = 2.0
messages = ['404923', '404925']
nosy_count = 6.0
nosy_names = ['paul.moore', 'tim.golden', 'zach.ware', 'eryksun', 'steve.dower', 'grv87']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue45597'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

### Linked PRs
* gh-155396
* gh-156611
* gh-156612
* gh-156613

**Comment by grv87** [comment]
@eryksun, are you proposing a change to discuss, or…?
Your code works for me. Although I found the comment misleading.

**Comment by eryksun** [comment]
> are you proposing a change to discuss, or…?

Yes, the suggested change in `ntpath.realpath()` will always keep the "\\\\?\\" prefix if the resolved path is not on the "UNC" device or a DOS drive. 

> Although I found the comment misleading.

[`GetFinalPathNameByHandleW()`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlew) supports getting the final path as `VOLUME_NAME_DOS` (i.e. using a DOS drive-letter volume name or a UNC share) or `VOLUME_NAME_GUID` (i.e. using a "Volume{GUID}" volume name). If a volume has not been assigned a DOS drive-letter name, then requesting `VOLUME_NAME_DOS` will fail. Python's implementation of `nt._getfinalpathname()` doesn't fall back on `VOLUME_NAME_GUID` in this case. It could, but it currently doesn't. But solving this issue by extending the implementation of `_getfinalpathname()` in "Modules/posixmodule.c" would make the solution overly difficult, and it still might miss edge cases. It's simpler to just change the implementation of `ntpath.realpath()` as mentioned above.

**Comment by eryksun** [comment]
Note that reproducing the issue requires the error after stripping the "\\\\?\\" prefix to match `initial_winerror`. So sometimes `realpath()` will work correctly, given the two error codes are different.

Also, there's no misbehavior if the path of the junction is registered with the mount point manager as the canonical DOS path of the volume, e.g. a junction created by "mountvol.exe" or WinAPI `SetVolumeMountPointW()` (requires administrator access). `nt._getfinalpathname()` returns the canonical DOS path.

> Why not open a PR?
> Or suggested changes must be discussed/approved/… before PR? 

Personally, I prefer to discuss ideas in the issue itself. It leaves a more accessible record. I've learned a lot from reading old issues. You're welcome to create a PR if you're eager to resolve this issue. I'll try to help with code review and testing.

> I haven't tested, but IIUC, these flags aren't fallbacks. They choose
> which variant should be returned when there are several possibilities
> (with drive letter or with volume GUID). But if there is no letter,
> `VOLUME_NAME_GUID` would give the same result as `VOLUME_NAME_DOS`.

If a volume has no canonical DOS mount point, then requesting `VOLUME_NAME_DOS` fails. In this case, I suggested that the implementation of `nt._getfinalpathname()` could automatically fall back on requesting the GUID name via `VOLUME_NAME_GUID`. This will succeed as long as the volume supports the mount point manager. In general, this would be a welcome enhancement. 

But, as I said, for this issue it's better to modify `realpath()` to never try removing the "\\\\?\\" prefix if the path doesn't start with a  drive-letter drive or "UNC" drive. 

> I meant, the term "drive path" is not wide-accepted.
> I'd just say `Regular paths start from \\?\<drive letter>:\`.

How about clarifying the comments as "For UNC drives, the path starts with \\\\?\\UNC\\", and  "For drive-letter drives, the path starts with \\\\?\\<drive letter>:\\"?

**Comment by grv87** [comment]
> If a volume has no canonical DOS mount point, then requesting `VOLUME_NAME_DOS` fails. In this case, I suggested that the implementation of `nt._getfinalpathname()` could automatically fall back on requesting the GUID name via `VOLUME_NAME_GUID`

I see. My mistake.
When `VOLUME_NAME_DOS` fails, Python does in `_getfinalpathname_nonstrict` what `VOLUME_NAME_GUID` would do, doesn't it?

> How about clarifying the comments as
👍
